sw/cheri/tests/i2c_tests.hh

```diff
@@ -0,0 +1,285 @@
+    0x6E, 0x73, 0x65, 0x20, 0x48, 0x41, 0x54, 0x7F, // rry PiSense HAT.
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // ................
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // ................
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // ................
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 // ................
+};
+
+/**
+ * The location of the Raspberry Pi Sense HAT's Accelerometer and
+ * Gyroscope's `WHO_AM_I` register, and the value that we expect
+ * this to report.
+ */
+constexpr uint8_t RpiHatAccelGyroWhoAmIAddr = 0x6A;
+constexpr uint8_t ExpectedRpiHatAccelGyroID = 0x68;
+
+/**
+ * The location of the Raspberry Pi Sense HAT's Magnetic Sensor's
+ * `WHO_AM_I` register, and the value that we expect this to report.
+ */
+constexpr uint8_t RpiHatIdMagneticWhoAmIAddr = 0x1C;
+constexpr uint8_t ExpectedRpiHatMagneticID = 0x3D;
+
+/**
+ * The location of the AS6212 Temperature Sensor in I2C.
+ */
+constexpr uint8_t As6212Addr = 0x48;
+
+/**
+ * The addresses of various I2C device registers used in testing.
+ */
+constexpr uint8_t RpiHatWhoAmIRegAddr = 0x0F;
+constexpr uint8_t As6212TempRegAddr = 0x00;
+constexpr uint8_t As6212ConfigRegAddr = 0x01;
+
+bool i2c_interrupt_flag = false;
+
+/**
+ * This test expects a Raspberry Pi Sense HAT to be connected to the Sonata
+ * board. The test configures the I2C, and then performs a write and read
+ * to the appropriate RPI HAT ID EEPROM pins to access the data stored in
+ * the I2C EEPROM header. Depending on the definition of
+ * `I2C_TEST_RPI_HAT_ID_EEPROM_CONTENTS`, this test will either verify that
+ * *something* is read, or will verify the exact contents match.
+ * Returns the number of failures that occur.
+ */
+int i2c_rpi_hat_id_eeprom_test(Capability<volatile OpenTitanI2c> i2c) {
+  int failures = 0;
+
+  // Setup the i2c bus, configuring it in host mode with speed of 100.
+  i2c->reset_fifos();
+  i2c->host_mode_set();
+  i2c->speed_set(100);
+
+  // Send two 0x0000 byte addresses and skip the STOP condition.
+  const uint8_t addr[] = {0, 0};
+  i2c->blocking_write(RpiHatIdEepromAddr, addr, sizeof(addr), true);
+
+  // Fill a read buffer with dummy data so that we see changes.
+  constexpr uint8_t DummyReadVal = 0x3D;
+  uint8_t eeprom_data[RpiHatIdEepromHeaderSize];
+  for (auto &read_val : eeprom_data) {
+    read_val = DummyReadVal;
+  }
+
+  // Read a page from the RPI Hat ID EEPROM
+  if (!i2c->blocking_read(RpiHatIdEepromAddr, eeprom_data,
+                          RpiHatIdEepromHeaderSize)) {
+    failures++;
+  }
+
+  // Verify the contents of the EEPROM header that were read.
+  bool buffer_changed = false;
+  for (uint32_t index = 0; index < RpiHatIdEepromHeaderSize; index++) {
```

Comment:
Use iterator?

Reply:
Keeping this because I'm simultaneously iterating over two arrays. I could perhaps get an iterator for both arrays and iterate through each in a loop, but for simplicity I'm just keeping a simple index-based loop here.